packages/atchops/src/base64.c

```diff
@@ -2,12 +2,22 @@
 #include <mbedtls/base64.h>
 #include <stddef.h>
 
-int atchops_base64_encode(const unsigned char *src, const size_t srclen, unsigned char *dst,
-                          const size_t dstsize, size_t *dstlen) {
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+int atchops_base64_encode(const unsigned char *src, const size_t srclen, unsigned char *dst, const size_t dstsize,
+                          size_t *dstlen) {
   return mbedtls_base64_encode(dst, dstsize, dstlen, src, srclen);
 }
 
-int atchops_base64_decode(const unsigned char *src, const size_t srclen, unsigned char *dst,
-                          const size_t dstsize, size_t *dstlen) {
+int atchops_base64_decode(const unsigned char *src, const size_t srclen, unsigned char *dst, const size_t dstsize,
+                          size_t *dstlen) {
   return mbedtls_base64_decode(dst, dstsize, dstlen, src, srclen);
 }
+
+size_t atchops_base64_encoded_size(const size_t plaintextsize) {
+  return MAX((size_t)(plaintextsize * 1.5 + 0.5), 8); // add 0.5 to ceiling
+}
+
+size_t atchops_base64_decoded_size(const size_t encodedsize) {
+  return MAX((size_t)(encodedsize * 0.75 + 0.5), 8); // add 0.5 to ceiling
+}
```

Comment:
I think this needs to be `* 3 / 2` and `* 3 / 4` to benefit from the compiler optimizations that we discussed earlier, but there are more caveats related to endianness and how to efficiently address the ceiling portion.

It should be possible to perform both of these functions entirely as bit wise operations. But it is too much of a time sink now, this is something we should revisit.

Just so I remember for later:

Let p be the plaintext size.
Let x be the first bit from the little end of P (do an AND against 0b00000001).
Let y be p bit shifted towards the little end (essentially removing the x bit from p, also ensuring that the new bit is zeroed)

Then x + y + p should be the optimized version of `plaintextsize * 1.5 + 0.5`

---
A similar algorithm can be created for the decoded_size function, but computations must be done twice to essentially get a half and a quarter of the encoded size to add together. A bit wise OR of the two `x` bits from this algorithm can be added to the final result to perform the rounding.

Reply:
Again, this is complete overkill, so for now, maybe change to:
`* 3 /2 + 1` and `*3/4+1` we may have a single bit of extra overhead, but that one bit will not hurt.

Reply:
```suggestion
size_t atchops_base64_encoded_size(const size_t plaintextsize) {
  return MAX((size_t)(plaintextsize * 3 / 2 + 1), 8); // add 0.5 to ceiling
}

size_t atchops_base64_decoded_size(const size_t encodedsize) {
  return MAX((size_t)(encodedsize * 3 / 4 + 1), 8); // add 0.5 to ceiling
}
```